sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -36,29 +36,41 @@ import org.apache.spark.sql.internal.StaticSQLConf._
 import org.apache.spark.status.ElementTrackingStore
 import org.apache.spark.util.{MutableURLClassLoader, Utils}
 
-
 /**
  * A class that holds all state shared across sessions in a given [[SQLContext]].
  */
-private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
+private[sql] class SharedState(val sparkContext: SparkContext, initConfig: Map[String, String])
+  extends Logging {
+  private val conf = sparkContext.conf.clone()
+  private val hadoopConf = new Configuration(sparkContext.hadoopConfiguration)
```

Comment:
we can even do:


```scala
private val (conf, hadoopConf) = {
  val clonedConf = sparkContext.conf.clone()
  val clonedHadoopConf = new Configuration(sparkContext.hadoopConfiguration)
  // If `SparkSession` is instantiated using an existing `SparkContext` instance and no existing
  // `SharedState`, all `SparkSession` level configurations have higher priority to generate a
  // `SharedState` instance. This will be done only once then shared across `SparkSession`s
  // See SPARK-26794.
  initConfig.foreach { case (k, v) =>
    logDebug(s"Applying initiate SparkSession options to SparkConf/HadoopConf: $k -> $v")
    conf.set(k, v)
    hadoopConf.set(k, v)
  }
  logDebug("Applied all initiate SparkSession options to the brand new SharedState")
  (clonedConf, clonedHadoopConf)
}
```

Reply:
this is more code, I think the current one is good enough. These 2 configs are mutable and we don't need to pretend that they are not.